### PR TITLE
Spillable columnar batch takes ownership and improve code coverage

### PR DESCRIPTION
**Pull request by abellina**
Closes https://github.com/NVIDIA/spark-rapids/issues/8669

This PR was originally done to address https://github.com/NVIDIA/spark-rapids/issues/8669 but I then spent time looking at code coverage of the code I had changed. I used both unit tests and integration tests to figure out that some classes were not being covered. 

I adjusted an integration test in hash_aggregate_test.py (`test_hash_grpby_sum_count_action`) to help me cover the code that would have instantitated `SpillableColumnarBatch`. In the process of doing so I found a bug in `AbstractProjectSplitIterator` that I am fixing here. As a result of testing this I did make changes to `RapidsConf` and the test framework that are marked as POC because I'd like feedback if we like the approach, if so I can spend time propagating the changes to the unit tests, or filing issues to do so. 

In a nutshell I think we need some changes to inject parameters, not unlike mocks, so that we can get the integration tests and also the unit tests to touch certain code paths that wouldn't be touched otherwise: in this example I forced the `GpuKeyBatchingIterator` to get started in the integration tests by setting `batchSizeBytes` to be 1, and I also forced the `PreProjectSplitIterator` to run by overriding the value of `getSplitUntilSize` which it used to trigger logic that wasn't covered.

### CODE REVIEW

**Comment by abellina** [comment]
The build will fail due to some style issues and lack of signature. I'll fix that if folks think the approach for the property override looks good, otherwise I'll need to rework things in order to be able to cover this code with tests (likely via unit tests and mocks)

File: integration_tests/src/main/python/spark_session.py [hunk 1]
```diff
@@ -71,10 +71,14 @@ def _set_all_confs(conf):
         _spark.conf.set("spark.rapids.sql.test.injectRetryOOM", "true")
     else:
         _spark.conf.set("spark.rapids.sql.test.injectRetryOOM", "false")
+    _spark.conf.set("spark.rapids.sql.test.isTestRun", "true")
```
Comment (revans2) [inline integration_tests/src/main/python/spark_session.py:74]: why do we need a new config? We already have a testing config.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala [hunk 2]
```diff
@@ -22,10 +22,11 @@ import scala.collection.mutable.ArrayBuffer
 import scala.util.control.NonFatal
 
 import ai.rapids.cudf._
-
 import org.apache.spark.{SparkEnv, TaskContext}
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:25]: The spaces between the imports looks like they need to be fixed.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala [hunk 3]
```diff
@@ -63,8 +64,30 @@ object GpuDeviceManager extends Logging {
 
   @volatile private var poolSizeLimit = 0L
 
+  private class DefaultOverrideProvider {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:67]: I don't see a lot of value in having a class for this.

```
def getSplitUntilSize: Long = {
  val conf = new RapidsConf(SQLConf.get)
  val override = if (conf.isTestEnabled) {
    conf.splitUntilSizeOverride
  } else {
    None
  }
  override.getOrElse(Math.max(poolSizeLimit / 8, 100 * 1024 * 1024))
}
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 4]
```diff
@@ -1332,6 +1332,14 @@ object RapidsConf {
     .booleanConf
     .createWithDefault(false)
 
+  val IS_TEST_RUN = conf("spark.rapids.sql.test.isTestRun")
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:1335]: Again, not sure why we want a new separate config from TEST_CONF

File: integration_tests/src/main/python/spark_session.py (hunk 1 above)
Comment (abellina) [reply integration_tests/src/main/python/spark_session.py:74]: @revans2 and I discussed this one offline. This was an optimization to not instantiate the `RapidsConf` for each batch. We discussed changing this so we can memoize the config per task instead, so I should be able to remove this config since the overhead should be much less.

Note that the other test config we have is: `spark.rapids.sql.test.enabled` which is more around verifying that all operators in a SQL query are on the GPU, so we assert that we have replaced everything and else the tests should fail. That said, less configs is better so I am all for simplifying this.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala (hunk 3 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:67]: I'll use your approach for this PR. I think we can the idea in the back of our heads if we find ourselves adding overrides all over the place so we can trigger special cased code, so we have a class that groups them (perhaps near RapidsConf).

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
@revans2 @jlowe this one is ready for another look. Thanks!

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala [hunk 5]
```diff
@@ -344,9 +344,14 @@ case class GpuOutOfCoreSortIterator(
...
       withResource(sortedTbl.contiguousSplit(splitIndexes: _*)) { splits =>
+        var currentSplit = 0
         val stillPending = if (hasFullySortedData) {
-          val sp = SpillableColumnarBatch(splits.head, sorter.projectedBatchTypes,
+          val ct = splits(currentSplit)
+          splits(currentSplit) = null
+          val sp = SpillableColumnarBatch(ct,
+            sorter.projectedBatchTypes,
             SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
+          currentSplit = currentSplit + 1
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:354]:
```suggestion
          currentSplit += 1
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala [hunk 6]
```diff
@@ -358,6 +363,8 @@ case class GpuOutOfCoreSortIterator(
           closeOnExcept(pendingObs) { _ =>
             stillPending.zip(lowerBoundaries).foreach {
               case (ct: ContiguousTable, lower: UnsafeRow) =>
+                splits(currentSplit) = null
+                currentSplit = currentSplit + 1
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:367]:
```suggestion
                currentSplit += 1
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/SpillableColumnarBatch.scala [hunk 7]
```diff
@@ -158,12 +158,14 @@ object SpillableColumnarBatch {
...
-        ct.getRowCount.toInt,
-        sparkTypes)
+    withResource(ct) { _ =>
+      val handle = RapidsBufferCatalog.addContiguousTable(ct, priority)
+      withResource(RapidsBufferCatalog.acquireBuffer(handle)) { _ =>
+        new SpillableColumnarBatchImpl(
+          handle,
+          ct.getRowCount.toInt,
+          sparkTypes)
+      }
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/SpillableColumnarBatch.scala:168]: Doesn't seem necessary to acquire the buffer since we're ignoring the result anyway?
```suggestion
      new SpillableColumnarBatchImpl(handle, ct.getRowCount.toInt, sparkTypes)
```

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/SpillableColumnarBatch.scala:168]: will change

**Review by revans2 — APPROVED** [review]

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
@jlowe I think I handled the comments you had so far.

**Review by jlowe — APPROVED** [review]
